server/src/main/java/org/elasticsearch/index/snapshots/blobstore/SlicedInputStream.java

```diff
@@ -148,12 +172,16 @@ public void reset() throws IOException {
                     throw new IOException("Mark has not been set");
                 }
 
-                // We do not call the SlicedInputStream's skipNBytes but call skipNBytes directly on the returned stream, to ensure that
-                // the skip is performed on the marked slice and no other slices are involved. This may help uncover any bugs.
                 nextSlice = markedSlice;
-                final InputStream stream = nextStream();
-                if (stream != null) {
-                    stream.skipNBytes(markedSliceOffset);
+                initialized = true;
+                IOUtils.close(currentStream);
+                if (nextSlice < numSlices) {
+                    currentStream = openSlice(nextSlice++);
+                    // We do not call the SlicedInputStream's skipNBytes but call skipNBytes directly on the returned stream, to ensure that
+                    // the skip is performed on the marked slice and no other slices are involved. This may help uncover any bugs.
+                    currentStream.skipNBytes(markedSliceOffset);
+                } else {
+                    currentStream = null;
```

Comment:
Can this `else` branch happen? Isn't `markedSlice < numSlices` an invariant?

Reply:
It can happen in the case one marks after reaching EOF. Added a test and made sure it works.